File: web/js/peacoq-ts/src/feedback.ts
import { FeedbackContent } from "./feedback-content";

export class Feedback {
  // TODO: give this a less lame type
  editOrState: string;
  editOrStateId: number;
  feedbackContent: FeedbackContent;
  routeId: number;
  constructor(f) {
    switch (f[0].tag) {
      case "State":
        this.editOrState = "state";
        break;
      case "Edit":
        this.editOrState = "edit";
        break;
      default:
        throw "Feedback tag was neither State nor Edit";
    };
    this.editOrStateId = f[0].contents;
    this.feedbackContent = FeedbackContent.create(f[1]);
    this.routeId = f[2];
  }
  toString() {
    return (
      "Feedback(" + this.editOrState + ", " + this.editOrStateId + ", " +
      this.feedbackContent + ", " + this.routeId + ")"
    );
  }
}

File: src/lib/Coq/Feedback.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

module Coq.Feedback where

import Data.Aeson            (ToJSON)
import Data.Proxy
import GHC.Generics          (Generic)
import Text.XML.Stream.Parse (tagName)

import Coq.EditId
import Coq.FeedbackContent
import Coq.StateId
import FromXML
import Utils

data EditOrStateId
  = Edit EditId
  | State StateId
  deriving (Generic, Show)

instance ToJSON EditOrStateId

type RouteId = Int

data Feedback
  = MkFeedback EditOrStateId FeedbackContent RouteId
  deriving (Generic, Show)

instance ToJSON Feedback

instance FromXML Feedback where
  instanceName Proxy = "Feedback"
  parseXML =
    tagName "feedback"
    (do
         object <- unpackRequireAttr "object"
         route <- castRequireAttr "route"
         return (object, route))
    $ \ (object, route) -> do
      editOrStateId <- case object of
        "edit" -> Edit <$> forceXML
        "state" -> State <$> forceXML
        _ -> error "feedback object was neither edit nor state"
      feedbackContent <- forceXML
      return $ MkFeedback editOrStateId feedbackContent route
